windows/devices/uwb/UwbDevice.cxx

```diff
@@ -191,12 +42,72 @@ UwbDevice::Initialize()
         FILE_FLAG_OVERLAPPED,
         nullptr));
 
-    // TODO: wil::unique_hfile handleDriver(CreateFile(m_deviceName))
+    wil::unique_hfile handleDriverNotifications(CreateFileA(
+        m_deviceName.c_str(),
+        GENERIC_READ | GENERIC_WRITE,
+        FILE_SHARE_READ | FILE_SHARE_WRITE,
+        nullptr,
+        OPEN_EXISTING,
+        FILE_ATTRIBUTE_NORMAL,
+        nullptr));
+
     // TODO: call CM_Register_Notification(handleDriver, filterType=CM_NOTIFY_FILTER_TYPE_DEVICEHANDLE)
     //   - handle CM_NOTIFY_ACTION_DEVICEQUERYREMOVE -> close handleDriver since device removal is requested
     //   - handle CM_NOTIFY_ACTION_DEVICEQUERYREMOVEFAILED -> query removal failed
 
     m_handleDriver = std::move(handleDriver);
+    m_handleDriverNotifications = std::move(handleDriverNotifications);
+    m_notificationThread = std::jthread([this] {
+        HandleNotifications();
+    });
+}
+
+void
+UwbDevice::HandleNotifications()
+{
+    for (;;) {
+        // TODO: check exit condition
+
+        // Determine the amount of memory required for the UWB_NOTIFICATION_DATA from the driver.
+        DWORD bytesRequired = 0;
+        BOOL ioResult = DeviceIoControl(m_handleDriver.get(), IOCTL_UWB_NOTIFICATION, nullptr, 0, nullptr, 0, &bytesRequired, nullptr);
+        if (!LOG_IF_WIN32_BOOL_FALSE(ioResult)) {
+            HRESULT hr = HRESULT_FROM_WIN32(GetLastError());
+            PLOG_ERROR << "error determining output buffer size for UWB notification, hr=" << std::showbase << std::hex << hr;
+            continue;
+        }
+
+        // Allocate memory for the UWB_NOTIFICATION_DATA structure, and pass this to the driver request.
+        DWORD uwbNotificationDataSize = bytesRequired;
+        std::vector<uint8_t> uwbNotificationDataBuffer(uwbNotificationDataSize);
+        ioResult = DeviceIoControl(m_handleDriver.get(), IOCTL_UWB_NOTIFICATION, nullptr, 0, std::data(uwbNotificationDataBuffer), uwbNotificationDataSize, &bytesRequired, nullptr);
+        if (!LOG_IF_WIN32_BOOL_FALSE(ioResult)) {
+            HRESULT hr = HRESULT_FROM_WIN32(GetLastError());
+            PLOG_ERROR << "error obtaining UWB notification buffer, hr=" << std::showbase << std::hex << hr << std::endl;
```

Comment:
I think PLog automatically prints each log message on its own line, so we don't need to append std::endl

Reply:
Yep, thanks, will remove.

Reply:
Removed.